Migrate `sin` and `sin_` from the TH to Aten (CUDA)

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28696 Remove definitions of acosh and asinh from TH
* #28527 Migrate `sinh` and `sinh_` from the TH to Aten (CUDA)
* #28482 Migrate `asin` and `asin_` from the TH to Aten (CUDA)
* **#28237 Migrate `sin` and `sin_` from the TH to Aten (CUDA)**

Benchmark (RHEL 7, gcc 8.3.1, P1000):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.sin(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.sin(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4649172620011086
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.4616892600006395
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5166665920005471
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5376560490003612
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6207812359989475
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.873208982999131
```

After:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4796977340010926
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.48329569199995603
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5380683220009814
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5299932739999349
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6144487999990815
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.8838113630008593
```


Close #24627

Differential Revision: [D18089072](https://our.internmc.facebook.com/intern/diff/D18089072)

xuhdev committed:
Migrate `sin` and `sin_` from the TH to Aten (CUDA)

Close #24627

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Close #24627

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Close #24627

[ghstack-poisoned]

ifedan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2486,15 +2486,12 @@
   dispatch:
-    CPU: _sin_out_cpu
-    CUDA: _sin_out_cuda
+    CPU: sin_out
```
we don't need this section, because both names are same

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Close #24627

[ghstack-poisoned]

ifedan commented:
Provide, please, performance metrics for this function before and after your change.

xuhdev commented:
Benchmark added

ifedan commented:
> Benchmark added. Seems the timing is pretty unstable though; I got numbers that are way different after a while. Also see benchmark on the previous two PRs on this stack. But this should be sufficient to show that there is likely no negligence in losing performance.

did you use?
OMP_NUM_THREADS=1
MKL_NUM_THREADS=1

xuhdev commented:
> did you use?
> OMP_NUM_THREADS=1
> MKL_NUM_THREADS=1

I didn't enable OMP (`USE_OPENMP=0`) and didn't have MKL installed.

xuhdev commented:
@ifedan I ended up benchmarking on a different machine which I have more control over the hardware. The results are pretty close and stable now. Please see my update in the benchmark results.

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Benchmark (RHEL 7, gcc 8.3.1, P1000):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.sin(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.sin(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4649172620011086
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.4616892600006395
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5166665920005471
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5376560490003612
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6207812359989475
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.873208982999131
```

After:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4796977340010926
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.48329569199995603
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5380683220009814
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5299932739999349
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6144487999990815
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.8838113630008593
```


Close #24627

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Benchmark (RHEL 7, gcc 8.3.1, P1000):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.sin(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.sin(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4649172620011086
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.4616892600006395
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5166665920005471
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5376560490003612
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6207812359989475
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.873208982999131
```

After:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4796977340010926
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.48329569199995603
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5380683220009814
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5299932739999349
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6144487999990815
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.8838113630008593
```


Close #24627

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate `sin` and `sin_` from the TH to Aten (CUDA)"


Benchmark (RHEL 7, gcc 8.3.1, P1000):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.sin(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.sin(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4649172620011086
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.4616892600006395
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5166665920005471
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5376560490003612
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6207812359989475
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.873208982999131
```

After:

```
torch.sin(a) a.numel() == 10000 for 20000 times torch.half
0.4796977340010926
torch.sin(a) a.numel() == 10000 for 20000 times torch.float
0.48329569199995603
torch.sin(a) a.numel() == 10000 for 20000 times torch.double
0.5380683220009814
torch.sin(a) a.numel() == 100000 for 20000 times torch.half
0.5299932739999349
torch.sin(a) a.numel() == 100000 for 20000 times torch.float
0.6144487999990815
torch.sin(a) a.numel() == 100000 for 20000 times torch.double
1.8838113630008593
```


Close #24627

Differential Revision: [D18089072](https://our.internmc.facebook.com/intern/diff/D18089072)

[ghstack-poisoned]
